### DISCUSSION TITLE
How to Load and Call a Function from a Lua File

### DISCUSSION THREAD

**Discussion by ocean-vivaforever** [gudzpoz/luajava#207] [Q&A]
Hello, I am working on porting a project that depends on the oldest version of luajava (www.keplerproject.org/luajava/) to the current version of luajava. I have two issues:
1. The Kepler luajava provides an LdoFile feature that allows directly loading .lua files. I would like to know what the corresponding implementation is in the new luajava. Is it ClassPathLoader?
2. I have a Lua file that defines a function, and I want to load this Lua file and call the function. How should I proceed?

The Lua file source code is as follows (from the luajava project test case):
```

function test()
    if true then
        io.write("test\n")
    end
end
for i=1,10 do
    io.write("test"..i.."\n")
end
io.stdout:flush()

```

In the oldest version of Kepler luajava, I only needed to call the following code:
```
LuaState L = LuaStateFactory.newLuaState();
L.openLibs();
L.LdoFile("luajava/simpleLuaFile");
LuaObject func= L.getLuaObject("test");
func.call(new Object[] { });
```
In the current version of luajava, I tried the following code:
```
try (Lua L = new LuaJit()) {
    L.openLibraries();
    ClassPathLoader classPathLoader = new ClassPathLoader();
    L.setExternalLoader(classPathLoader);
    Buffer buffer = classPathLoader.load("luajava/simpleLuaFile", null);
    L.load(buffer, "luajava/simpleLuaFile");
    L.getGlobal("test");
    LuaValue function = L.get();
    System.out.println(function.type());
}
```
this just return NIL

please help

**Comment by ocean-vivaforever** [comment]
find the problem
just use L.run() instead of L.load()
```java
try (Lua L = new LuaJit()) {
    L.openLibraries();
    ClassPathLoader classPathLoader = new ClassPathLoader();
    L.setExternalLoader(classPathLoader);
    Buffer buffer = classPathLoader.load("luajava/simpleLuaFile", null);
    L.run(buffer, "luajava/simpleLuaFile");
    L.getGlobal("test");
    LuaValue function = L.get();
    funcion.call();
    System.out.println(function.type());
}
```

**Comment by gudzpoz** [comment]
Yes, `Lua.load` eventually uses `lua_load`, and it `only loads a chunk; it does not run it` (https://www.lua.org/manual/5.1/manual.html#lua_load). I guess I just assume people are familiar with Lua C API and I will probably need to add more documentation to the `Lua.load` methods. Thanks

(By the way, you can save a few lines of code with the following:
```java
try (Lua L = new LuaJit()) {
    L.openLibraries();
    L.setExternalLoader(new ClassPathLoader());
    L.run("require('luajava.simpleLuaFile')"); // I guess I will add a Lua.require to facilitate usages like this
    LuaValue function = L.get("test");
    funcion.call();
    System.out.println(function.type());
}
```
)
